jkube-kit/build/service/buildpacks/src/test/java/org/eclipse/jkube/kit/service/buildpacks/AbstractBuildPackCliDownloaderTest.java

```diff
@@ -0,0 +1,240 @@
+      }
+
+      @Test
+      @DisplayName("warning is logged indicating that pack binary download failed")
+      void logWarningAboutDownloadFailure() {
+        // Given + When
+        assertThatIllegalStateException().isThrownBy(buildPackCliDownloader::getPackCLIIfPresentOrDownload);
+
+        // Then
+        ArgumentCaptor<String> downloadFailureMessage = ArgumentCaptor.forClass(String.class);
+        verify(kitLogger).warn(downloadFailureMessage.capture());
+        assertThat(downloadFailureMessage.getValue()).contains("Not able to download pack CLI : ");
+      }
+
+      @Test
+      @DisplayName("info is logged indicating we attempt to use a local pack binary as a fallback")
+      void logInfoIndicatingFallbackToLocalPackCli() {
+        // Given + When
+        assertThatIllegalStateException().isThrownBy(buildPackCliDownloader::getPackCLIIfPresentOrDownload);
+
+        // Then
+        verify(kitLogger).info("Checking for local pack CLI");
+      }
+
+      @Test
+      @DisplayName("local pack binary doesn't exist, then throw exception")
+      void givenNoLocalPackBinaryInUserPath_thenThrowException() {
+        // When + Then
+        assertThatIllegalStateException()
+            .isThrownBy(buildPackCliDownloader::getPackCLIIfPresentOrDownload)
+            .withMessage("No local pack binary found");
+      }
+
+      @Test
+      @DisplayName("local pack binary exists and is valid, then return local pack binary path")
+      void givenLocalPackCliExistsAndIsValid_thenReturnPathToLocalPackBinary() throws IOException {
+        // Given
+        givenPackCliPresentOnUserPath(String.format("/%s", getApplicablePackBinary()));
+
+        // When
+        File downloadedCli = buildPackCliDownloader.getPackCLIIfPresentOrDownload();
+
+        // Then
+        assertThat(downloadedCli).isEqualTo(temporaryFolder.toPath().resolve("bin").resolve(getApplicablePackBinary()).toFile());
+      }
+
+      @Test
+      @DisplayName("local pack binary exists but invalid, then throw exception")
+      void localPackCliCorrupt_thenThrowException() throws IOException {
+        // Given
+        givenPackCliPresentOnUserPath(String.format("/%s", getInvalidApplicablePackBinary()));
+
+        // When + Then
+        assertThatIllegalStateException()
+            .isThrownBy(buildPackCliDownloader::getPackCLIIfPresentOrDownload)
+            .withMessage("No local pack binary found");
+      }
+
+      private void givenPackCliPresentOnUserPath(String packResource) throws IOException {
+        File bin = new File(temporaryFolder, "bin");
+        File pack = new File(Objects.requireNonNull(getClass().getResource(packResource)).getFile());
+        Files.createDirectory(bin.toPath());
+        Files.copy(pack.toPath(), bin.toPath().resolve(pack.getName()), COPY_ATTRIBUTES);
+      }
+    }
+  }
+
+  private Properties createPackProperties(TestHttpStaticServer server) {
```

Comment:
nit: since this method isn't (or at least shouldn't) be reused, there's no point in hiding this below the entire suite and having the logic extracted. It'd be better to have it in the generic setup method

Reply:
I'll update it in a follow up PR

Reply:
I checked and this method seems to be getting reused

In global `@BeforeEach` for initializing downloader with valid artifact dir: 
https://github.com/eclipse/jkube/blob/a1fcb571333121eef45810487e046672c65e90c2/jkube-kit/build/service/buildpacks/src/test/java/org/eclipse/jkube/kit/service/buildpacks/AbstractBuildPackCliDownloaderTest.java#L69-L71

In `@BeforeEach` for DownloadFails scenario , we initialize downloader with invalid artifact dir: 

https://github.com/eclipse/jkube/blob/a1fcb571333121eef45810487e046672c65e90c2/jkube-kit/build/service/buildpacks/src/test/java/org/eclipse/jkube/kit/service/buildpacks/AbstractBuildPackCliDownloaderTest.java#L152-L154

Reply:
OK, so this reveals a problem that could be fixed and used to simplify the tests.

If properly configured, there should be no need to initialize two different servers.
A single server could serve the two kinds of artifacts. What you'd need to change is the URLs, which in turn should be done in the specific nested tests (this would also add more clarity).